pkg/build/generator/generator.go

```diff
@@ -0,0 +1,243 @@
+	}
+	build.Labels[buildapi.BuildConfigLabel] = bcCopy.Name
+
+	for originalImage, newImage := range imageSubstitutions {
+		glog.Infof("Substituting %s for %s", newImage, originalImage)
+		substituteImageReferences(build, originalImage, newImage)
+	}
+	return build, nil
+}
+
+// substituteImageReferences replaces references to an image with a new value
+func substituteImageReferences(build *buildapi.Build, oldImage string, newImage string) {
+	switch {
+	case build.Parameters.Strategy.Type == buildapi.DockerBuildStrategyType &&
+		build.Parameters.Strategy.DockerStrategy != nil &&
+		build.Parameters.Strategy.DockerStrategy.Image == oldImage:
+		build.Parameters.Strategy.DockerStrategy.Image = newImage
+	case build.Parameters.Strategy.Type == buildapi.STIBuildStrategyType &&
+		build.Parameters.Strategy.STIStrategy != nil &&
+		build.Parameters.Strategy.STIStrategy.Image == oldImage:
+		build.Parameters.Strategy.STIStrategy.Image = newImage
+	case build.Parameters.Strategy.Type == buildapi.CustomBuildStrategyType:
+		// update env variable references to the old image with the new image
+		strategy := build.Parameters.Strategy.CustomStrategy
+		if strategy.Env == nil {
+			strategy.Env = make([]kapi.EnvVar, 1)
+			strategy.Env[0] = kapi.EnvVar{Name: buildapi.CustomBuildStrategyBaseImageKey, Value: newImage}
+		} else {
+			found := false
+			for i := range strategy.Env {
+				glog.Infof("Checking env variable %s %s", strategy.Env[i].Name, strategy.Env[i].Value)
+				if strategy.Env[i].Name == buildapi.CustomBuildStrategyBaseImageKey {
+					found = true
+					if strategy.Env[i].Value == oldImage {
+						strategy.Env[i].Value = newImage
+						glog.Infof("Updated env variable %s %s", strategy.Env[i].Name, strategy.Env[i].Value)
+						break
+					}
+				}
+			}
+			if !found {
+				strategy.Env = append(strategy.Env, kapi.EnvVar{Name: buildapi.CustomBuildStrategyBaseImageKey, Value: newImage})
+			}
+		}
+		// update the actual custom build image with the new image, if applicable
+		if strategy.Image == oldImage {
+			strategy.Image = newImage
+		}
+	}
+}
+
+// getNextBuildName returns name of the next build and increments BuildConfig's LastVersion.
+func getNextBuildName(bc *buildapi.BuildConfig) string {
+	bc.LastVersion++
+	return fmt.Sprintf("%s-%d", bc.Name, bc.LastVersion)
+}
+
+// generateBuildFromBuild creates a new build based on a given Build.
+func generateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
+	obj, _ := kapi.Scheme.Copy(build)
+	buildCopy := obj.(*buildapi.Build)
+	return &buildapi.Build{
+		Parameters: buildCopy.Parameters,
+		ObjectMeta: kapi.ObjectMeta{
+			Name:   getNextBuildNameFromBuild(buildCopy),
+			Labels: buildCopy.ObjectMeta.Labels,
+		},
+	}
+}
+
+// getNextBuildNameFromBuild returns name of the next build with random uuid added at the end
+func getNextBuildNameFromBuild(build *buildapi.Build) string {
+	return fmt.Sprintf("%s-%d", build.Name, int32(util.Now().Unix()))
```

Comment:
if i generate a build from a build that was generated from a build, i'm going to end up with a mess.  you should strip the previous timestamp(if present) from the buildname when generating a build from a build.